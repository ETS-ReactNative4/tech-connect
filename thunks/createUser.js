import { isLoading, hasErrored, logInUser } from '../actions'

export const createUser = (user) => {

  const url = 'https://tech-connect-be.herokuapp.com/api/v1/users'
  return async (dispatch) => {
    try {
      dispatch(isLoading(true))
      const response = await fetch(url, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json'
        },
        body: JSON.stringify(user)
      })

      if (!response.ok) {
        const errorText = await response.json()
        throw Error(errorText.error)
      }

      dispatch(isLoading(false))
      const newUser = await response.json()
      const loggedInUser = {
        email: user.email,
        id: newUser.data.id,
        api_key: newUser.data.attributes.api_key,
      }
      dispatch(logInUser(loggedInUser))
      dispatch(hasErrored(''))
    } catch (err) {
      dispatch(hasErrored(err.message))
    }
  }
}

