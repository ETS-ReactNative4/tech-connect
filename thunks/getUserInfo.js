export const getUserInfo = async (id, apiKey) => {

  const url = `https://tech-connect-be.herokuapp.com/api/v1/users/${id}?api_key=${apiKey}`
  const key = {api_key: apiKey}
  try {
    const response = await fetch(url)
    if (!response.ok) {
      const errorText = await response.json()
      throw Error(errorText.error)
    }
    
    const userInfo = await response.json()
    const cleanedUser = {id: userInfo.data.id, ...userInfo.data.attributes}
    return cleanedUser
  } catch(err) {
    console.log(err)
  }
}