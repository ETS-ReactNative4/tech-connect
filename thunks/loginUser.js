import { isLoading, hasErrored, logInUser } from '../actions'

export const loginUser = (user) => {
  const url = 'https://tech-connect-be.herokuapp.com/api/v1/login'
  return async (dispatch) => {
    try {
      dispatch(isLoading(true))
      const response = await fetch(url, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json'
        },
        body: JSON.stringify(user)
      })

      if (!response.ok) {
        const errorText = await response.json()
        throw Error(errorText.error)
      }

      dispatch(isLoading(false))
      const currentUser = await response.json();
      const loggedInUser = {
        email: user.email,
        id: currentUser.data.id,
        api_key: currentUser.data.attributes.api_key,
      }
      const userObject = {...loggedInUser,...currentUser.data.attributes}
      dispatch(logInUser(userObject))
      dispatch(hasErrored(''))
    } catch(err) {
      dispatch(hasErrored(err.message))
    }
  }
}

