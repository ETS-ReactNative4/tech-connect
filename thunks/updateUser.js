import { isLoading, hasErrored, updateUserAction } from '../actions'

export const updateUser = (user) => {
  const url = 'https://tech-connect-be.herokuapp.com/api/v1/users'
  return async (dispatch) => {

    try {
      dispatch(isLoading(true))
      const response = await fetch(url, {
        method: 'PATCH',
        headers: {
          'Content-Type': 'application/json'
        },
        body: JSON.stringify(user)
      })

      if(!response.ok) {
        const errorText = await response.json()
        throw Error(errorText.error)
      }

      dispatch(isLoading(false))
      const updatedUser = await response.json()
      dispatch(updateUserAction(updatedUser.data.attributes))
      dispatch(hasErrored(''))
    } catch(err) {
      dispatch(hasErrored(err.message))
    }
  }
}