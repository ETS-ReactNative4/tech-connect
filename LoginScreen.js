import React from 'react';
import { StyleSheet, Text, View, TextInput, Button, TouchableHighlight, TouchableOpacity } from 'react-native';
import { connect } from 'react-redux'
import { LinearGradient } from 'expo'
import { Input } from 'react-native-elements';
import { createUser } from './thunks/createUser'
import { loginUser } from './thunks/loginUser'
import Icon from 'react-native-vector-icons/EvilIcons';


export class LoginScreen extends React.Component {
  constructor(props) {
    super(props)
    this.state = {
      register: false,
      email: '',
      password: '',
      password_confirmation: '',
      error: false
    }
  }

  toggleRegister = () => {
    this.setState({
      register: !this.state.register,
      error: false
    })
  }

  handleSubmit = async () => {
    const { email, password, password_confirmation } = this.state
    const createInfo = { email, password, password_confirmation }
    const loginInfo = { email, password }

    if (this.state.register) {
      await this.props.createUser(createInfo)
      this.handleUpdateProfile()
    } else {
      await this.props.loginUser(loginInfo)
      this.handleUpdateProfile()
    }

    if(this.props.error) {
      this.setState({ error: true })
    }
  }
  

  handleUpdateProfile = () => {
    if (this.props.error === '' && this.state.register) {
      this.props.navigation.navigate('EditProfile')
    } else if (this.props.error === '') {
      this.props.navigation.navigate('Home')
    }
  }

  render() {
    const btnText = this.state.register ? 'REGISTER' : 'SIGN IN'
    const toggleText = this.state.register ? 'Sign In' : 'Register'
    const errorMessage = this.state.error ? this.props.error : null

    return (
      <View style={styles.container}>
        <Text style={styles.header}>Tech</Text>
        <Text style={[styles.header, {marginBottom: 50}]}>Connect</Text>
        <Input
          leftIcon= {
            <LinearGradient
                colors={['#4AA9C5', '#6364c1', '#93548F']}
                start={0.4}
                style={styles.gradient} >
              <Icon
                raised={true}
                name='envelope'
                size={28}
                color='#FFF'
              />
            </LinearGradient>
            }
          leftIconContainerStyle={styles.icon} 
          inputContainerStyle={styles.inputContainer} 
          containerStyle={styles.input} 
          placeholder="Email" style={styles.input} 
          name='email' 
          onChangeText={(text) => this.setState({email: text})} 
        />
        <Input 
          leftIcon= {
            <LinearGradient
              colors={['#4AA9C5', '#6364c1', '#93548F']}
              start={0.4}
              style={styles.gradient} 
            >
              <Icon
                raised={true}
                name='lock'
                size={32}
                color='#FFF'
              />
            </LinearGradient>
          }
          leftIconContainerStyle={styles.icon} 
          inputContainerStyle={styles.inputContainer} 
          containerStyle={styles.input} 
          secureTextEntry={true} 
          name='password' 
          placeholder="Password" 
          onChangeText={(text) => this.setState({password: text})} 
        /> 
        {
          this.state.register && 
            <Input 
              leftIcon= {
                <LinearGradient
                  colors={['#4AA9C5', '#6364c1', '#93548F']}
                  start={0.4}
                  style={styles.gradient} >
                  <Icon
                    raised={true}
                    name='lock'
                    size={32}
                    color='#FFF'
                  />
                </LinearGradient>
              }
              leftIconContainerStyle={styles.icon} 
              inputContainerStyle={styles.inputContainer} 
              containerStyle={styles.input} 
              secureTextEntry={true} 
              placeholder="Confirm Password" 
              onChangeText={(text) => this.setState({password_confirmation: text})}
            /> 
        }
        <Text style={styles.error}>{ errorMessage }</Text>
        <TouchableHighlight style={styles.button}>
          <Button title={btnText} color='#E8FDFF' onPress={this.handleSubmit}/>
        </TouchableHighlight>
        <View style={styles.textContainer}>
          <TouchableOpacity onPress={this.toggleRegister} >
            <Text style={[styles.text, {textDecorationLine: 'underline', fontWeight: 'bold'}]}>{toggleText}</Text>
          </TouchableOpacity>
          <Text style={styles.text} > if you do not have an account</Text>
        </View>
      </View>
    );
  }
}


export const mapStateToProps = (state) => ({
  user: state.user,
  isLoading: state.isLoading,
  error: state.error
})

export const mapDispatchToProps = (dispatch) => ({
  createUser: (user) => dispatch(createUser(user)),
  loginUser: (user) => dispatch(loginUser(user))
})

export default connect(mapStateToProps, mapDispatchToProps)(LoginScreen)


const styles = StyleSheet.create({

  container: {
    flex: 1,
    backgroundColor: '#4AA9C5',
    alignItems: 'center',
    justifyContent: 'center',
  },
  header: {
    height: 75,
    alignItems: 'center',
    color: '#fff',
    marginTop: 0,
    marginBottom: 0,
    fontSize: 90,
    fontFamily: 'American Typewriter',
    lineHeight: 90,
  },
  gradient: {
    borderRadius: 300, 
    width: 48, 
    height: 48, 
    justifyContent: 'center', 
    alignItems: 'center'
  },
  text: {
    color: '#E8FDFF',
  },
  icon: {
    marginRight: 10,
    marginLeft: -5,
    width: 48,
    height: 48,
    borderRadius: 50,
  },
  error: {
    color: 'red',
    marginBottom: 3
  },
  inputContainer: {
    borderBottomWidth: 0,
    height: 40
  },
  input: {
    backgroundColor: '#E8FDFF',
    borderRadius: 30,
    padding: 10,
    width: 300,
    margin: 10
  },
  button: {
    justifyContent: 'center', 
    alignItems: 'center',
    backgroundColor: '#8FC9DB',
    borderRadius: 25,
    height: 44,
    width: 140,
    shadowOffset: {  width: 2,  height: 2},
    shadowRadius: 5,
    shadowColor: 'black',
    shadowOpacity: .2
  },
  textContainer: {
    display: 'flex',
    flexDirection: 'row',
    marginTop: 8
  }
});