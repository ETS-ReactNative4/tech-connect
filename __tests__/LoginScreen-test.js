import React from 'react'
import { LoginScreen, mapStateToProps, mapDispatchToProps } from '../LoginScreen'
import { Input } from 'react-native-elements';
import { shallow } from 'enzyme'
import { createUser } from '../thunks/createUser'
import { loginUser } from '../thunks/loginUser'


jest.mock('../thunks/createUser')
jest.mock('../thunks/loginUser')

import renderer from 'react-test-renderer'

describe('LoginScreen', () => {
  let wrapper
  let navigation

  beforeEach(() => {
    navigation = { navigate: jest.fn() };
    onPressEvent = jest.fn()
    wrapper = shallow(<LoginScreen 
      loginUser={jest.fn} 
      createUser={jest.fn()} 
      navigation={navigation} 
      error={''}
    />)
  })

  it('renders the snapshot', () => {
    const tree = renderer.create(
      <LoginScreen />
      ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('toggles the screen between register and sign in', () => {
    wrapper.instance().toggleRegister()
    expect(wrapper.instance().state.register).toEqual(true)
  })

  it('calls handleSubmit on button press', () => {
    const spy = jest.spyOn(wrapper.instance(), 'handleSubmit')
    wrapper.instance().forceUpdate()
    wrapper.find('Button').simulate('press')
    expect(spy).toHaveBeenCalled();
  })

  it('creates a new user account', () => {
    wrapper.setState({ register: true })
    mockUser = {user: 'Kylie', password: 'hello', password_confirmation: 'hello' }
    wrapper.instance().handleSubmit()
    expect(wrapper.instance().props.createUser(mockUser))
  })

  it('signs in a current user', () => {
    wrapper.setState({ register: false })
    mockUser = {user: 'Kylie', password: 'hello' }
    wrapper.instance().handleSubmit()
    expect(wrapper.instance().props.loginUser(mockUser))
  })

  it('calls handleUpdateProfile', async () => {
    const spy = jest.spyOn(wrapper.instance(), 'handleUpdateProfile')
    await wrapper.instance().handleSubmit()
    expect(spy).toHaveBeenCalled()
  })

  it('should navigate to the Profile screen', () => {
    wrapper.setState({ register: true })
    wrapper.instance().handleUpdateProfile()
    expect(wrapper.instance().props.navigation.navigate).toHaveBeenCalledWith('EditProfile')
  })

  it('should navigate to the Home screen', () => {
    wrapper.instance().handleUpdateProfile()
    expect(wrapper.instance().props.navigation.navigate).toHaveBeenCalledWith('Home')
  })

  it('should update the email on input', () => {
    wrapper.find(Input).at(0).simulate('changeText', 'email')
    expect(wrapper.instance().state.email).toEqual('email')
  })

  it('should set the error in state to true if the error prop is true', async () => {
    wrapper = shallow(<LoginScreen loginUser={jest.fn} error={'error'} />)
    await wrapper.instance().handleSubmit()
    expect(wrapper.instance().state.error).toEqual(true)
  })

  it('should update the password on input', () => {
    wrapper.find(Input).at(1).simulate('changeText', 'password')
    expect(wrapper.instance().state.password).toEqual('password')
  })

  it('should update the cofirmed password on input', () => {
    wrapper.instance().setState({register: true})
    wrapper.find(Input).at(2).simulate('changeText', 'password')
    expect(wrapper.instance().state.password_confirmation).toEqual('password')
  })

  describe('mapStateToProps', () => {
    it('should return a user object', () => {
      const mockState = {
        user: { name: 'Kylie' },
        connections: { name: 'Bob' },
        isLoading: false
      }
      const expected = { user: mockState.user, isLoading: false }
      const result = mapStateToProps(mockState)
      expect(result).toEqual(expected)
    })
  })

  describe('mapDispatchToProps', () => {
    it('should call dispatch with the correct params for createUser', () => {
      const mockDispatch = jest.fn()
      const url = 'create.com'
      const actionToDispatch = createUser(url)
      const result = mapDispatchToProps(mockDispatch)

      result.createUser(url)

      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch)
    })

    it('should call dispatch with the correct params for loginUser', () => {
      const mockDispatch = jest.fn()
      const url = 'login.com'
      const actionToDispatch = loginUser(url)
      const result = mapDispatchToProps(mockDispatch)

      result.loginUser(url)

      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch)
    })
  })
})