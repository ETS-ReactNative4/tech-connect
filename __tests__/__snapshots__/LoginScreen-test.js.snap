// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginScreen renders the snapshot 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#4AA9C5",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <Text>
    Tech Connect
  </Text>
  <View
    style={
      Object {
        "backgroundColor": "#E8FDFF",
        "borderRadius": 30,
        "margin": 10,
        "padding": 10,
        "paddingHorizontal": 10,
        "width": 300,
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "borderBottomWidth": 0,
          "borderColor": "#86939e",
          "flexDirection": "row",
          "height": 40,
          "transform": Array [
            Object {
              "translateX": 0,
            },
          ],
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "pink",
            "borderRadius": 50,
            "height": 48,
            "justifyContent": "center",
            "marginLeft": -5,
            "marginRight": 10,
            "width": 48,
          }
        }
      >
        <ExponentLinearGradient
          colors={
            Array [
              4283083205,
              4284703937,
              4287845519,
            ]
          }
          startPoint={
            Array [
              undefined,
              undefined,
            ]
          }
          style={
            Object {
              "alignItems": "center",
              "borderRadius": 300,
              "height": 48,
              "justifyContent": "center",
              "width": 48,
            }
          }
        >
          <Text />
        </ExponentLinearGradient>
      </View>
      <TextInput
        allowFontScaling={true}
        name="email"
        onChangeText={[Function]}
        placeholder="Email"
        style={
          Object {
            "alignSelf": "center",
            "color": "black",
            "flex": 1,
            "fontSize": 18,
            "minHeight": 40,
          }
        }
        testID="RNE__Input__text-input"
        underlineColorAndroid="transparent"
      />
    </View>
  </View>
  <View
    style={
      Object {
        "backgroundColor": "#E8FDFF",
        "borderRadius": 30,
        "margin": 10,
        "padding": 10,
        "paddingHorizontal": 10,
        "width": 300,
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "borderBottomWidth": 0,
          "borderColor": "#86939e",
          "flexDirection": "row",
          "height": 40,
          "transform": Array [
            Object {
              "translateX": 0,
            },
          ],
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "pink",
            "borderRadius": 50,
            "height": 48,
            "justifyContent": "center",
            "marginLeft": -5,
            "marginRight": 10,
            "width": 48,
          }
        }
      >
        <ExponentLinearGradient
          colors={
            Array [
              4283083205,
              4284703937,
              4287845519,
            ]
          }
          startPoint={
            Array [
              undefined,
              undefined,
            ]
          }
          style={
            Object {
              "alignItems": "center",
              "borderRadius": 300,
              "height": 48,
              "justifyContent": "center",
              "width": 48,
            }
          }
        >
          <Text />
        </ExponentLinearGradient>
      </View>
      <TextInput
        allowFontScaling={true}
        name="password"
        onChangeText={[Function]}
        placeholder="Password"
        secureTextEntry={true}
        style={
          Object {
            "alignSelf": "center",
            "color": "black",
            "flex": 1,
            "fontSize": 18,
            "minHeight": 40,
          }
        }
        testID="RNE__Input__text-input"
        underlineColorAndroid="transparent"
      />
    </View>
  </View>
  <Text
    style={
      Object {
        "color": "red",
        "marginBottom": 3,
      }
    }
  />
  <View
    accessible={true}
    isTVSelectable={true}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#8FC9DB",
        "borderRadius": 25,
        "height": 44,
        "justifyContent": "center",
        "shadowColor": "black",
        "shadowOffset": Object {
          "height": 2,
          "width": 2,
        },
        "shadowOpacity": 0.2,
        "shadowRadius": 5,
        "width": 140,
      }
    }
  >
    <View
      accessibilityRole="button"
      accessibilityStates={Array []}
      accessible={true}
      isTVSelectable={true}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <View
        style={
          Array [
            Object {},
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#007AFF",
                "fontSize": 18,
                "padding": 8,
                "textAlign": "center",
              },
              Object {
                "color": "#E8FDFF",
              },
            ]
          }
        >
          SIGN IN
        </Text>
      </View>
    </View>
  </View>
  <View
    style={
      Object {
        "display": "flex",
        "flexDirection": "row",
        "marginTop": 8,
      }
    }
  >
    <View
      accessible={true}
      isTVSelectable={true}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#E8FDFF",
            },
            Object {
              "fontWeight": "bold",
              "textDecorationLine": "underline",
            },
          ]
        }
      >
        Register
      </Text>
    </View>
    <Text
      style={
        Object {
          "color": "#E8FDFF",
        }
      }
    >
       if you do not have an account
    </Text>
  </View>
</View>
`;
